qiskit/transpiler/passes/synthesis/hls_plugins.py

```diff
@@ -0,0 +1,924 @@
+        if coupling_map is None or qubits is None:
+            # The abstract synthesis uses a fully connected coupling map, allowing
+            # arbitrary connections between qubits.
+            used_coupling_map = CouplingMap.from_full(len(pattern))
+        else:
+            # The concrete synthesis uses the coupling map restricted to the set of
+            # qubits over which the permutation gate is defined. If we allow using other
+            # qubits in the coupling map, replacing the node in the DAGCircuit that
+            # defines this PermutationGate by the DAG corresponding to the constructed
+            # decomposition becomes problematic. Note that we allow the reduced
+            # coupling map to be disconnected.
+            used_coupling_map = coupling_map.reduce(qubits, check_if_connected=False)
+
+        graph = used_coupling_map.graph.to_undirected()
+        swapper = ApproximateTokenSwapper(graph, seed=seed)
+
+        try:
+            swapper_result = swapper.map(
+                pattern_as_dict, trials, parallel_threshold=parallel_threshold
+            )
+        except rx.InvalidMapping:
+            swapper_result = None
+
+        if swapper_result is not None:
+            decomposition = QuantumCircuit(len(graph.node_indices()))
+            for swap in swapper_result:
+                decomposition.swap(*swap)
+            return decomposition
+
+        return None
+
+
+class MCXSynthesisNDirtyI15(HighLevelSynthesisPlugin):
+    r"""Synthesis plugin for a multi-controlled X gate based on [1].
+
+    This plugin name is :``mcx.n_dirty_i15`` which can be used as the key on
+    an :class:`~.HLSConfig` object to use this method with :class:`~.HighLevelSynthesis`.
+
+    For a multi-controlled X gate with :math:`k\ge 3` control qubits this synthesis
+    method requires :math:`k - 2` additional dirty auxiliary qubits. The synthesized
+    circuit consists of :math:`2 * k - 1` qubits and at most :math:`8 * k - 6` CX gates.
+
+    The plugin supports the following plugin-specific options:
+
+    * num_clean_ancillas: The number of clean auxiliary qubits available.
+    * num_dirty_ancillas: The number of dirty auxiliary qubits available.
+    * relative_phase: When set to ``True``, the method applies the optimized multi-controlled
+      X gate up to a relative phase, in a way that, by lemma 8 of [1], the relative
+      phases of the ``action part`` cancel out with the phases of the ``reset part``.
+    * action_only: when set to ``True``, the method applies only the ``action part``
+      of lemma 8 of [1].
+
+    References:
+        1. Iten et. al., *Quantum Circuits for Isometries*, Phys. Rev. A 93, 032318 (2016),
+           `arXiv:1501.06911 <http://arxiv.org/abs/1501.06911>`_
+    """
+
+    def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **options):
+        """Run synthesis for the given MCX gate."""
+
+        if high_level_object.name != "mcx":
```

Comment:
This shouldn't be necessary since the HLS should only give us things called `mcx` right? The other plugins also don't have this additional safety 🙂

Reply:
This is a good point and I can remove these checks as in the normal flow only the `HighLevelSynthesis` transpiler pass would call the plugins and the interface guarantees that the names would be correct. 

However, I am slightly worried about the case when a user might decide to call synthesis plugins directly (as they are documented in the API docs) on MCX gates, and in particular on something like a `CCXGate` (which in Qiskit is not an `MCXGate` but this decisions seems to have been rather arbitrary). So I wanted the plugins to complain in this case. Of course, the same problem exists if a user would called a QFT-synthesis plugin on a Clifford object, however intuitively it's harder to confuse `QFT` with `Clifford` than `CCX` with `MCX`. What do you think?